buffer/ring_growing.go

```diff
@@ -16,57 +16,72 @@ limitations under the License.
 
 package buffer
 
+import "sync/atomic"
+
 // RingGrowing is a growing ring buffer.
 // Not thread safe.
 type RingGrowing struct {
 	data     []interface{}
-	n        int // Size of Data
-	beg      int // First available element
-	readable int // Number of data items available
+	n        int32 // Size of Data
+	beg      int32 // First available element
+	readable int32 // Number of data items available
 }
 
 // NewRingGrowing constructs a new RingGrowing instance with provided parameters.
 func NewRingGrowing(initialSize int) *RingGrowing {
 	return &RingGrowing{
 		data: make([]interface{}, initialSize),
-		n:    initialSize,
+		n:    int32(initialSize),
 	}
 }
 
 // ReadOne reads (consumes) first item from the buffer if it is available, otherwise returns false.
 func (r *RingGrowing) ReadOne() (data interface{}, ok bool) {
-	if r.readable == 0 {
+	oldReadable := atomic.LoadInt32(&r.readable)
+	oldN := atomic.LoadInt32(&r.n)
+	if oldReadable == 0 {
+		return nil, false
+	}
+	if !atomic.CompareAndSwapInt32(&r.readable, oldReadable, oldReadable-1) {
 		return nil, false
 	}
-	r.readable--
-	element := r.data[r.beg]
-	r.data[r.beg] = nil // Remove reference to the object to help GC
-	if r.beg == r.n-1 {
-		// Was the last element
-		r.beg = 0
+	oldBeg := atomic.LoadInt32(&r.beg)
+
+	if oldBeg == oldN-1 {
+		if !atomic.CompareAndSwapInt32(&r.beg, oldBeg, 0) {
+			return nil, false
+		}
 	} else {
-		r.beg++
+		if !atomic.CompareAndSwapInt32(&r.beg, oldBeg, oldBeg+1) {
+			return nil, false
+		}
 	}
+	element := r.data[oldBeg]
+	r.data[oldBeg] = nil // Remove reference to the object to help GC
 	return element, true
 }
 
 // WriteOne adds an item to the end of the buffer, growing it if it is full.
 func (r *RingGrowing) WriteOne(data interface{}) {
-	if r.readable == r.n {
+	oldReadable := atomic.LoadInt32(&r.readable)
+	oldN := atomic.LoadInt32(&r.n)
+	oldBeg := atomic.LoadInt32(&r.beg)
+
+	if oldN == oldReadable {
 		// Time to grow
-		newN := r.n * 2
+		newN := oldN * 2
 		newData := make([]interface{}, newN)
-		to := r.beg + r.readable
+		to := oldBeg + oldReadable
 		if to <= r.n {
 			copy(newData, r.data[r.beg:to])
 		} else {
 			copied := copy(newData, r.data[r.beg:])
 			copy(newData[copied:], r.data[:(to%r.n)])
 		}
-		r.beg = 0
+		atomic.StoreInt32(&r.n, newN)
 		r.data = newData
-		r.n = newN
+		atomic.StoreInt32(&r.beg, 0)
 	}
 	r.data[(r.readable+r.beg)%r.n] = data
-	r.readable++
+	atomic.StoreInt32(&r.readable, oldReadable+1)
```

Comment:
I'm not sure about the use case of `RingGrowing`, but for general cases (multi-producer and multi-consumer) this is still not thread safe.

It's possible that two goroutines execute line `66-68` at the same time thus get the same `oldReadable`. They may try to execute `atomic.StoreInt32` with exactly the same `oldReadable` which means we lost one item that should stay in the ring.

Lock-free unbounded queue is not so easy to get right. I haven't been doing this style of programming for a long time. This is the [last paper](https://www.cs.tau.ac.il/~mad/publications/ppopp2013-x86queues.pdf) I read about concurrent queues.

Reply:
Yes, it's also not thread safe enough in this circumstance.

I think should we just simple use a atomic integer to replace default int for r.beg and check the margin before use it as a index for slice?

Reply:
As I don't know the intended use case of `RingGrowing`, I don't know if it's appropriate to change this to a thread safe implementation.

If changing this to a thread safe version is the way to go, I'd suggest just use locks for simplicity. If performance is indeed a problem, consider use some proven implementation of concurrent queue.